### PR TITLE
Initial e2e tests

### PR DESCRIPTION
**Pull request by kalantar**
Together these test:
(a) experiment with uniform strategy where baseline, candidate are defined before experiment
(b) experiment with uniform strategy where baseline defined before experiment, candidate after
(c) experiment in (b) but paused and resumed

### CODE REVIEW

File: test/e2e/e2e-scenario-0c.sh [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+header "Create $NAMESPACE gateway and vs"
+kubectl apply -n $NAMESPACE -f $DIR/../../doc/tutorials/istio/bookinfo/bookinfo-gateway.yaml
+kubectl get gateway -n $NAMESPACE
+kubectl get vs -n $NAMESPACE
+
+#header "Test productpage via service IP"
+#IP=`kubectl -n $NAMESPACE get services | grep productpage | awk '{print $3}'`
+#PORT=`kubectl -n $NAMESPACE get services | grep productpage | awk '{print $5}' | awk -F/ '{print $1}'`
+#echo "Bookinfo is accessed at $IP:$PORT"
+#curl -H "Host: bookinfo.example.com" -Is "http://$IP:$PORT/productpage"
```
Comment (kimikowang) [inline test/e2e/e2e-scenario-0c.sh:47]: Should these be deleted?

File: test/e2e/e2e-scenario-0c.sh [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+# kubectl -n $NAMESPACE get experiments.iter8.tools
+
+# # force experiment complete
+# # verify experiment complete
+# header "rollforward experiment"
+# kubectl -n $NAMESPACE patch experiments.iter8.tools $EXPERIMENT --type='merge' \
+#   -p='{"spec": {"manualOverride": { "action": "terminate", "trafficSplit": { "reviews-v2": 0, "reviews-v3": 100 }}}}'
+# sleep 2
+# test_experiment_status $EXPERIMENT "IterationUpdate: Iteration"
+# kubectl -n $NAMESPACE get experiments.iter8.tools "$EXPERIMENT-restart" -o yaml
```
Comment (kimikowang) [inline test/e2e/e2e-scenario-0c.sh:126]: Should these be deleted?

File: test/e2e/e2e-scenario-0b.sh [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+header "Create $NAMESPACE gateway and vs"
+kubectl apply -n $NAMESPACE -f $DIR/../../doc/tutorials/istio/bookinfo/bookinfo-gateway.yaml
+kubectl get gateway -n $NAMESPACE
+kubectl get vs -n $NAMESPACE
+
+#header "Test productpage via service IP"
+#IP=`kubectl -n $NAMESPACE get services | grep productpage | awk '{print $3}'`
+#PORT=`kubectl -n $NAMESPACE get services | grep productpage | awk '{print $5}' | awk -F/ '{print $1}'`
+#echo "Bookinfo is accessed at $IP:$PORT"
+#curl -H "Host: bookinfo.example.com" -Is "http://$IP:$PORT/productpage"
```
Comment (kimikowang) [inline test/e2e/e2e-scenario-0b.sh:43]: Should these be deleted?

File: test/e2e/e2e-scenario-0a.sh [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+header "Start iter8 end-to-end testing"
+
+header "Scenario 0a - pre-existing deployments - uniform"
+
+header "Set Up"
+
+header "Clean Up Any Existing"
+# delete any existing experiment with same name
+kubectl -n $NAMESPACE delete experiment $EXPERIMENT --ignore-not-found
```
Comment (huang195) [inline test/e2e/e2e-scenario-0a.sh:21]: Any reason that we are parameterize EXPERIMENT but not deployment 2 lines down?

File: test/e2e/e2e-scenario-0a.sh [hunk 5]
```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env bash
+
+# Exit on error
+#set -e
+
+DIR="$( cd "$( dirname "$0" )" >/dev/null 2>&1; pwd -P )"
+source "$DIR/library.sh"
+
+NAMESPACE="${NAMESPACE:-bookinfo-iter8}"
```
Comment (huang195) [inline test/e2e/e2e-scenario-0a.sh:9]: These parameters are used for running these testing scripts outside of Travis?

File: test/e2e/library.sh [hunk 6]
```diff
@@ -14,3 +14,19 @@ function make_banner() {
...
+  echo "expected = $expected"
+  echo "  actual = $actual"
+
+  if [[ "$actual" != *"$expected"* ]]; then
+    echo "Experiment $experiment failed unexpectedly"
+    echo " expected status message: $expected"
+    echo "      got status message: $actual"
+    exit 1
+  fi
+}
```
Comment (huang195) [inline test/e2e/library.sh:32]: You might want to echo something when experiment result is of the expected value

File: test/e2e/e2e-scenario-0c.sh (hunk 1 above)
Comment (kalantar) [reply test/e2e/e2e-scenario-0c.sh]: yes. done.

File: test/e2e/e2e-scenario-0c.sh (hunk 2 above)
Comment (kalantar) [reply test/e2e/e2e-scenario-0c.sh:120]: no. these are additional tests I want to add. However, they are not fully tested. need to switch focus to kubecon scenarios.

File: test/e2e/e2e-scenario-0b.sh (hunk 3 above)
Comment (kalantar) [reply test/e2e/e2e-scenario-0b.sh]: yes. done.

File: test/e2e/e2e-scenario-0a.sh (hunk 4 above)
Comment (kalantar) [reply test/e2e/e2e-scenario-0a.sh:21]: Not really. Actually, the only parameter I used was IP when testing against an IKS cluster. However, I prefer to use a variable so I don't make a spelling mistake.

Deployment is used twice with different values so not sure it meets the bar for variable.

File: test/e2e/e2e-scenario-0a.sh (hunk 5 above)
Comment (kalantar) [reply test/e2e/e2e-scenario-0a.sh:9]: yes.

File: test/e2e/library.sh (hunk 6 above)
Comment (kalantar) [reply test/e2e/library.sh:34]: good idea. done.

**Comment by kalantar** [comment]
Any remaining comments @kimikowang Please note that there are a couple of new minor changes to the controller that got mixed in as I tried to get it working (and fix the last_state problem at the same time).

**Pull request merged**